src/core/Events.js

```diff
@@ -96,102 +96,81 @@ L.Evented = L.Class.extend({
 		var typeListeners = this._events[type];
 		if (!typeListeners) {
 			typeListeners = {
-				listeners: {},
+				listeners: [],
 				count: 0
 			};
 			this._events[type] = typeListeners;
 		}
 
-		var contextId = context && context !== this && L.stamp(context),
-		    newListener = {fn: fn, ctx: context};
-
-		if (!contextId) {
-			contextId = 'no_context';
-			newListener.ctx = undefined;
-		}
-
-		// fn array for context
-		var listeners = typeListeners.listeners[contextId];
-		if (!listeners) {
-			listeners = [];
-			typeListeners.listeners[contextId] = listeners;
+		if (context === this) {
+			// Less memory footprint.
+			context = undefined;
 		}
+		var newListener = {fn: fn, ctx: context},
+		    listeners = typeListeners.listeners;
 
 		// check if fn already there
 		for (var i = 0, len = listeners.length; i < len; i++) {
-			if (listeners[i].fn === fn) {
+			if (listeners[i].fn === fn && listeners[i].ctx === context) {
 				return;
 			}
 		}
 
-		listeners.push(newListener);
+		typeListeners.listeners.push(newListener);

```

Comment:
No good reason ;)
I removed `listeners` local variable at first, and then apparently forgot to put it back here.
Fixing it, thanks! :)